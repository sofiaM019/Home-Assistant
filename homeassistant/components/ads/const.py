"""Support for Automation Device Specification (ADS)."""

from __future__ import annotations

from enum import StrEnum
from typing import TYPE_CHECKING

from homeassistant.util.hass_dict import HassKey

if TYPE_CHECKING:
    from .hub import AdsHub

DOMAIN = "ads"

DATA_ADS: HassKey[AdsHub] = HassKey(DOMAIN)
STATE_KEY_STATE = "state"

SERVICE_WRITE_DATA_BY_NAME = "write_data_by_name"

CONF_ADS_VAR = "adsvar"
CONF_ADS_FACTOR = "factor"
CONF_ADS_TYPE = "adstype"
CONF_ADS_VALUE = "value"


class AdsType(StrEnum):
    """Supported Types."""

    BOOL = "bool"
    BYTE = "byte"
    INT = "int"
    UINT = "uint"
    SINT = "sint"
    USINT = "usint"
    DINT = "dint"
    UDINT = "udint"
    WORD = "word"
    DWORD = "dword"
    LREAL = "lreal"
    REAL = "real"
    STRING = "string"
    TIME = "time"
    DATE = "date"
    DATE_AND_TIME = "dt"
    TOD = "tod"
