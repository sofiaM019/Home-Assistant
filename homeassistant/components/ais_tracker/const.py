"""AIS tracker constants."""

DOMAIN = "ais_tracker"

CONF_MMSIS = "mmsis"
DEFAULT_PORT = 12346
