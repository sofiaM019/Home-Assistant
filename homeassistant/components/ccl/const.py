"""Constants for the CCL Electronics integration."""

DOMAIN = "ccl"
