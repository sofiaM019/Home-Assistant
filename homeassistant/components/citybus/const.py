"""CityBus constants"""

DOMAIN = "citybus"

CONF_ROUTE = "route"
CONF_DIRECTION = "direction"
CONF_STOP = "stop"
