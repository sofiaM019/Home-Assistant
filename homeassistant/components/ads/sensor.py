"""Support for ADS sensors."""

from __future__ import annotations

import voluptuous as vol

from homeassistant.components.sensor import (
    CONF_STATE_CLASS,
    DEVICE_CLASSES_SCHEMA as SENSOR_DEVICE_CLASSES_SCHEMA,
    PLATFORM_SCHEMA as SENSOR_PLATFORM_SCHEMA,
    STATE_CLASSES_SCHEMA as SENSOR_STATE_CLASSES_SCHEMA,
    SensorDeviceClass,
    SensorEntity,
    SensorStateClass,
)
from homeassistant.const import CONF_DEVICE_CLASS, CONF_NAME, CONF_UNIT_OF_MEASUREMENT
from homeassistant.core import HomeAssistant
import homeassistant.helpers.config_validation as cv
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType, StateType

from . import ADS_TYPEMAP, CONF_ADS_FACTOR, CONF_ADS_TYPE
from .const import CONF_ADS_VAR, DATA_ADS, STATE_KEY_STATE, AdsType
from .entity import AdsEntity
from .hub import AdsHub

DEFAULT_NAME = "ADS sensor"
CONF_OPTIONS = "options"

PLATFORM_SCHEMA = SENSOR_PLATFORM_SCHEMA.extend(
    {
        vol.Required(CONF_ADS_VAR): cv.string,
        vol.Optional(CONF_ADS_FACTOR): cv.positive_int,
        vol.Optional(CONF_ADS_TYPE, default=AdsType.INT): vol.All(
            vol.Coerce(AdsType),
            vol.In(
                [
                    AdsType.BOOL,
                    AdsType.BYTE,
                    AdsType.INT,
                    AdsType.UINT,
                    AdsType.SINT,
                    AdsType.USINT,
                    AdsType.DINT,
                    AdsType.UDINT,
                    AdsType.WORD,
                    AdsType.DWORD,
                    AdsType.LREAL,
                    AdsType.REAL,
                ]
            ),
        ),
        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
        vol.Optional(CONF_DEVICE_CLASS): SENSOR_DEVICE_CLASSES_SCHEMA,
        vol.Optional(CONF_STATE_CLASS): SENSOR_STATE_CLASSES_SCHEMA,
        vol.Optional(CONF_UNIT_OF_MEASUREMENT): cv.string,
        vol.Optional(CONF_OPTIONS): vol.All(cv.ensure_list, [cv.string]),
    }
)


def setup_platform(
    hass: HomeAssistant,
    config: ConfigType,
    add_entities: AddEntitiesCallback,
    discovery_info: DiscoveryInfoType | None = None,
) -> None:
    """Set up an ADS sensor device."""
    ads_hub = hass.data[DATA_ADS]

    ads_var: str = config[CONF_ADS_VAR]
    ads_type: AdsType = config[CONF_ADS_TYPE]
    name: str = config[CONF_NAME]
    factor: int | None = config.get(CONF_ADS_FACTOR)
    device_class: SensorDeviceClass | None = config.get(CONF_DEVICE_CLASS)
    state_class: SensorStateClass | None = config.get(CONF_STATE_CLASS)
    unit_of_measurement: str | None = config.get(CONF_UNIT_OF_MEASUREMENT)
    options: list[str] | None = config.get(CONF_OPTIONS, [])

    if device_class == SensorDeviceClass.ENUM and not options:
        raise ValueError("Options must be provided for ENUM device class")

    entity = AdsSensor(
        ads_hub,
        ads_var,
        ads_type,
        name,
        factor,
        device_class,
        state_class,
        unit_of_measurement,
        options,
    )

    add_entities([entity])


class AdsSensor(AdsEntity, SensorEntity):
    """Representation of an ADS sensor entity."""

    def __init__(
        self,
        ads_hub: AdsHub,
        ads_var: str,
        ads_type: AdsType,
        name: str,
        factor: int | None,
        device_class: SensorDeviceClass | None,
        state_class: SensorStateClass | None,
        unit_of_measurement: str | None,
        options: list[str] | None,
    ) -> None:
        """Initialize AdsSensor entity."""
        super().__init__(ads_hub, name, ads_var)
        self._ads_type = ads_type
        self._factor = factor
        self._attr_device_class = device_class
        self._attr_state_class = state_class
        self._attr_native_unit_of_measurement = unit_of_measurement
        self._options = options or []

    async def async_added_to_hass(self) -> None:
        """Register device notification."""
        await self.async_initialize_device(
            self._ads_var,
            ADS_TYPEMAP[self._ads_type],
            STATE_KEY_STATE,
            self._factor,
        )

    @property
    def native_value(self) -> StateType:
        """Return the state of the device."""
        state = self._state_dict.get(STATE_KEY_STATE)
        if self._attr_device_class == SensorDeviceClass.ENUM:
            if isinstance(state, int) and 0 <= state < len(self._options):
                return self._options[state]
            return None
        return state
