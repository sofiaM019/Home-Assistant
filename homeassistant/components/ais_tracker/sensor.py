"""Sensors for AIS tracker."""

from collections.abc import Callable
from dataclasses import dataclass

from pyais.constants import ManeuverIndicator, NavigationStatus

from homeassistant.components.sensor import (
    SensorDeviceClass,
    SensorEntity,
    SensorEntityDescription,
)
from homeassistant.const import DEGREE, UnitOfSpeed
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import CONF_MMSIS, DOMAIN
from .coordinator import AisTrackerConfigEntry, AisTrackerCoordinator
from .entity import AistrackerEntity


@dataclass(frozen=True, kw_only=True)
class AisSensorEntityDescription(SensorEntityDescription):
    """A class that describes AEMET OpenData sensor entities."""

    keys: list[str] | None = None
    value_fn: Callable[[str | float], float | int | str | None] = lambda value: value


def _course_value(value: str | float) -> float | None:
    if float(value) == 360:
        return None
    return float(value)


def _heading_value(value: str | float) -> int | None:
    if int(value) == 511:
        return None
    return int(value)


def _turn_value(value: str | float) -> float | None:
    if float(value) == -128.0:
        return None
    return (float(value) / 4.733) ** 2


SENSORS: tuple[AisSensorEntityDescription, ...] = (
    AisSensorEntityDescription(
        key="course",
        translation_key="course",
        native_unit_of_measurement=DEGREE,
        value_fn=_course_value,
    ),
    AisSensorEntityDescription(
        key="heading",
        translation_key="heading",
        native_unit_of_measurement=DEGREE,
        value_fn=_heading_value,
    ),
    AisSensorEntityDescription(
        key="maneuver",
        translation_key="maneuver",
        device_class=SensorDeviceClass.ENUM,
        options=[status.name for status in ManeuverIndicator],
        value_fn=lambda value: ManeuverIndicator(int(value)).name,
    ),
    AisSensorEntityDescription(
        key="speed",
        device_class=SensorDeviceClass.SPEED,
        native_unit_of_measurement=UnitOfSpeed.KNOTS,
        value_fn=lambda value: float(value) / 10,
    ),
    AisSensorEntityDescription(
        key="status",
        translation_key="status",
        device_class=SensorDeviceClass.ENUM,
        options=[status.name for status in NavigationStatus],
        value_fn=lambda value: NavigationStatus(int(value)).name,
    ),
    AisSensorEntityDescription(
        key="turn",
        translation_key="turn",
        native_unit_of_measurement="Â°/min",
        value_fn=_turn_value,
    ),
)


async def async_setup_entry(
    hass: HomeAssistant,
    entry: AisTrackerConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up device tracker for AIS tracker."""
    coordinator = entry.runtime_data

    async_add_entities(
        AisTrackerSensorEntity(coordinator, mmsi, description)
        for mmsi in entry.data[CONF_MMSIS]
        for description in SENSORS
    )


class AisTrackerSensorEntity(AistrackerEntity, SensorEntity):
    """Represent a tracked device."""

    entity_description: AisSensorEntityDescription

    def __init__(
        self,
        coordinator: AisTrackerCoordinator,
        mmsi: str,
        description: AisSensorEntityDescription,
    ) -> None:
        """Set up AIS tracker tracker entity."""
        super().__init__(coordinator, mmsi)
        self.entity_description = description
        self._attr_unique_id = f"{DOMAIN}_{mmsi}_{description.key}"

    @property
    def native_value(self) -> float | int | str | None:
        """Return the state of the device."""
        if (data := self.data) is not None and (
            value := data.get(self.entity_description.key)
        ) is not None:
            return self.entity_description.value_fn(value)
        return None
