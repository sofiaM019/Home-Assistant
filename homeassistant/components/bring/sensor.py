"""Sensor platform for the Bring! integration."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from enum import StrEnum

from bring_api import BringUserSettingsResponse
from bring_api.const import BRING_SUPPORTED_LOCALES
from bring_api.types import BringList

from homeassistant.components.sensor import (
    DOMAIN as SENSOR_DOMAIN,
    SensorDeviceClass,
    SensorEntity,
    SensorEntityDescription,
)
from homeassistant.const import EntityCategory
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers import entity_registry as er
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.typing import StateType

from . import BringConfigEntry
from .const import DOMAIN, UNIT_ITEMS
from .coordinator import BringData, BringDataUpdateCoordinator
from .entity import BringBaseEntity
from .util import list_language, sum_attributes


@dataclass(kw_only=True, frozen=True)
class BringSensorEntityDescription(SensorEntityDescription):
    """Bring Sensor Description."""

    value_fn: Callable[[BringData, BringUserSettingsResponse], StateType]


class BringSensor(StrEnum):
    """Bring sensors."""

    URGENT = "urgent"
    CONVENIENT = "convenient"
    DISCOUNTED = "discounted"
    LIST_LANGUAGE = "list_language"
    LIST_ACCESS = "list_access"


SENSOR_DESCRIPTIONS: tuple[BringSensorEntityDescription, ...] = (
    BringSensorEntityDescription(
        key=BringSensor.URGENT,
        translation_key=BringSensor.URGENT,
        value_fn=lambda lst, _: sum_attributes(lst, "urgent"),
        native_unit_of_measurement=UNIT_ITEMS,
    ),
    BringSensorEntityDescription(
        key=BringSensor.CONVENIENT,
        translation_key=BringSensor.CONVENIENT,
        value_fn=lambda lst, _: sum_attributes(lst, "convenient"),
        native_unit_of_measurement=UNIT_ITEMS,
    ),
    BringSensorEntityDescription(
        key=BringSensor.DISCOUNTED,
        translation_key=BringSensor.DISCOUNTED,
        value_fn=lambda lst, _: sum_attributes(lst, "discounted"),
        native_unit_of_measurement=UNIT_ITEMS,
    ),
    BringSensorEntityDescription(
        key=BringSensor.LIST_LANGUAGE,
        translation_key=BringSensor.LIST_LANGUAGE,
        value_fn=(
            lambda lst, settings: x.lower()
            if (x := list_language(lst["uuid"], settings))
            else None
        ),
        entity_category=EntityCategory.DIAGNOSTIC,
        options=[x.lower() for x in BRING_SUPPORTED_LOCALES],
        device_class=SensorDeviceClass.ENUM,
    ),
    BringSensorEntityDescription(
        key=BringSensor.LIST_ACCESS,
        translation_key=BringSensor.LIST_ACCESS,
        value_fn=lambda lst, _: lst["status"].lower(),
        entity_category=EntityCategory.DIAGNOSTIC,
        options=["registered", "shared", "invitation"],
        device_class=SensorDeviceClass.ENUM,
    ),
)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: BringConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the sensor platform."""
    coordinator = config_entry.runtime_data
    lists_added: set[str] = set()

    @callback
    def add_entities() -> None:
        """Add or remove sensor entities."""
        nonlocal lists_added
        entities: list[BringSensorEntity] = []
        entity_registry = er.async_get(hass)

        for bring_list in coordinator.lists:
            if bring_list["listUuid"] not in lists_added:
                entities.extend(
                    BringSensorEntity(
                        coordinator,
                        bring_list,
                        description,
                    )
                    for description in SENSOR_DESCRIPTIONS
                )
            lists_added.add(bring_list["listUuid"])

        user = {x["listUuid"] for x in coordinator.user_settings["userlistsettings"]}
        for list_uuid in user | lists_added:
            if any(
                bring_list["listUuid"] == list_uuid for bring_list in coordinator.lists
            ):
                continue

            for description in SENSOR_DESCRIPTIONS:
                if entity_id := entity_registry.async_get_entity_id(
                    SENSOR_DOMAIN,
                    DOMAIN,
                    f"{coordinator.config_entry.unique_id}_{list_uuid}_{description.key}",
                ):
                    entity_registry.async_remove(entity_id)

            lists_added.discard(list_uuid)

        if entities:
            async_add_entities(entities)

    coordinator.async_add_listener(add_entities)
    add_entities()


class BringSensorEntity(BringBaseEntity, SensorEntity):
    """A sensor entity."""

    entity_description: BringSensorEntityDescription

    def __init__(
        self,
        coordinator: BringDataUpdateCoordinator,
        bring_list: BringList,
        entity_description: BringSensorEntityDescription,
    ) -> None:
        """Initialize the entity."""
        super().__init__(coordinator, bring_list)
        self.entity_description = entity_description
        self._attr_unique_id = f"{coordinator.config_entry.unique_id}_{self._list_uuid}_{self.entity_description.key}"

    @property
    def native_value(self) -> StateType:
        """Return the state of the sensor."""

        if list_data := next(
            filter(lambda x: x["uuid"] == self._list_uuid, self.coordinator.data), None
        ):
            return self.entity_description.value_fn(
                list_data,
                self.coordinator.user_settings,
            )
        return None
