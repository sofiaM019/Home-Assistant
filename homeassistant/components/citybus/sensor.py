"""CityBus sensor."""

import logging
from datetime import datetime

from typing import cast

from homeassistant.components.sensor import SensorDeviceClass, SensorEntity
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_NAME
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import CONF_ROUTE, CONF_DIRECTION, CONF_STOP, DOMAIN
from .coordinator import CityBusDataUpdateCoordinator

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(
    hass: HomeAssistant,
    config: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Load values from configuration and initialize the platform."""
    _LOGGER.debug(config.data)
    entry_route = config.data[CONF_ROUTE]
    entry_direction = config.data[CONF_DIRECTION]
    entry_stop = config.data[CONF_STOP]
    coordinator_key = f"{entry_route}-{entry_direction}-{entry_stop}"

    coordinator: CityBusDataUpdateCoordinator = hass.data[DOMAIN].get(coordinator_key)

    async_add_entities(
        (
            CityBusNextBusSensor(
                coordinator,
                cast(str, config.unique_id),
                config.data[CONF_ROUTE],
                config.data[CONF_DIRECTION],
                config.data[CONF_STOP],
                config.data.get(CONF_NAME) or config.title,
            ),
        ),
    )


class CityBusNextBusSensor(
    CoordinatorEntity[CityBusDataUpdateCoordinator], SensorEntity
):
    """Sensor class that displays upcoming CityBus times.

    To function, this requires knowing the key or code for the route, direction, and stop."""

    _attr_device_class = SensorDeviceClass.TIMESTAMP
    _attr_translation_key = "citybus"

    def __init__(
        self,
        coordinator: CityBusDataUpdateCoordinator,
        unique_id: str,
        route_key: str,
        direction_key: str,
        stop_code: str,
        name: str,
    ) -> None:
        """Initialize sensor with all required config."""
        super().__init__(coordinator)
        self.route_key = route_key
        self.direction_key = direction_key
        self.stop_code = stop_code
        self._attr_extra_state_attributes: dict[str, str] = {
            "route_key": route_key,
            "direction_key": direction_key,
            "stop_code": stop_code,
        }
        self._attr_unique_id = unique_id
        self._attr_name = name

    def _log_debug(self, message, *args):
        """Log debug message with prefix."""
        msg = f"{self.route_key}:{self.direction_key}:{self.stop_code}:{message}"
        _LOGGER.debug(msg, *args)

    def _log_err(self, message, *args):
        """Log error message with prefix."""
        msg = f"{self.route_key}:{self.direction_key}:{self.stop_code}:{message}"
        _LOGGER.error(msg, *args)

    async def async_added_to_hass(self) -> None:
        """Read data from coordinator after adding to hass."""
        self._handle_coordinator_update()
        await super().async_added_to_hass()

    @callback
    def _handle_coordinator_update(self) -> None:
        """Update sensor with new estimate times."""
        results = self.coordinator.get_estimate_data(
            self.route_key, self.direction_key, self.stop_code
        )

        self._log_debug("Estimate results: %s", results)

        if not results:
            self._log_err("Error getting estimates: %s", str(results))
            self._attr_native_value = None
            return

        estimates = results

        # Short circuit if there are no estimates available
        if not estimates:
            self._log_debug("No upcoming buses available")
            self._attr_native_value = None
            self._attr_extra_state_attributes["upcoming"] = "No upcoming estimates"
        else:
            self._attr_extra_state_attributes["upcoming"] = ", ".join(
                str(e["estimatedDepartTimeUtc"]) for e in estimates
            )

            first_bus = estimates[0]
            self._attr_native_value = datetime.fromisoformat(
                first_bus["estimatedDepartTimeUtc"]
            )

        self.async_write_ha_state()
