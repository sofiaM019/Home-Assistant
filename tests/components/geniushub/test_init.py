"""Tests for the Genius Hub component."""

from unittest.mock import AsyncMock

from syrupy import SnapshotAssertion

from homeassistant.components.geniushub import DOMAIN
from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
from homeassistant.const import CONF_MAC, CONF_TOKEN
from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr, entity_registry as er

from . import setup_integration

from tests.common import MockConfigEntry


async def test_device_info(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    mock_geniushub_cloud: AsyncMock,
    mock_cloud_config_entry: MockConfigEntry,
    device_registry: dr.DeviceRegistry,
) -> None:
    """Test device registry integration."""
    await setup_integration(hass, mock_cloud_config_entry)
    device_entry = device_registry.async_get_device(
        identifiers={(DOMAIN, mock_cloud_config_entry.entry_id)}
    )
    assert device_entry is not None
    assert device_entry == snapshot
    device_entry = device_registry.async_get_device(
        identifiers={(DOMAIN, f"{mock_cloud_config_entry.entry_id}_30")}
    )
    assert device_entry is not None
    assert device_entry == snapshot


async def test_cloud_unique_id_migration(
    hass: HomeAssistant,
    entity_registry: er.EntityRegistry,
    mock_geniushub_cloud: AsyncMock,
) -> None:
    """Test that the cloud unique ID is migrated to the entry_id."""
    entry = MockConfigEntry(
        domain=DOMAIN,
        title="Genius hub",
        data={
            CONF_TOKEN: "abcdef",
            CONF_MAC: "aa:bb:cc:dd:ee:ff",
        },
        entry_id="1234",
    )
    entry.add_to_hass(hass)
    entity_registry.async_get_or_create(
        SENSOR_DOMAIN, DOMAIN, "aa:bb:cc:dd:ee:ff_device_78", config_entry=entry
    )
    await hass.config_entries.async_setup(entry.entry_id)
    await hass.async_block_till_done()
    assert hass.states.get("sensor.geniushub_aa_bb_cc_dd_ee_ff_device_78")
    entity_entry = entity_registry.async_get(
        "sensor.geniushub_aa_bb_cc_dd_ee_ff_device_78"
    )
    assert entity_entry.unique_id == "1234_device_78"
