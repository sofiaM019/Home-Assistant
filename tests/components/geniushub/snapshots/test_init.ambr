# serializer version: 1
# name: test_device_info
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'geniushub',
        '01J71MQF0EC62D620DGYNG2R8H',
      ),
    }),
    'is_new': False,
    'labels': set({
    }),
    'manufacturer': None,
    'model': None,
    'model_id': None,
    'name': 'GeniusHub',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'suggested_area': None,
    'sw_version': None,
    'via_device_id': None,
  })
# ---
# name: test_device_info.1
  DeviceRegistryEntrySnapshot({
    'area_id': 'study',
    'config_entries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'geniushub',
        '01J71MQF0EC62D620DGYNG2R8H_30',
      ),
    }),
    'is_new': False,
    'labels': set({
    }),
    'manufacturer': None,
    'model': None,
    'model_id': None,
    'name': 'Study',
    'name_by_user': None,
    'primary_config_entry': <ANY>,
    'serial_number': None,
    'suggested_area': 'Study',
    'sw_version': None,
    'via_device_id': None,
  })
# ---
